### DISCUSSION TITLE
Help on migration

### DISCUSSION THREAD

**Discussion by govardha** [kannibalox/pyrosimple#50] [Q&A]
Hello,

I have run into a predicament and I am trying to figure out a course of action on migration from my existing node which is in a funky state(I was performing a routine patching and some libc++ module got corrupted & removed & now I am unable to ssh into the box).  The saving grace is I have a standalone ftp daemon which doesn't depend on this and I am able to backup what really matters to me in a new box which I was just building with pyrosimple being my primary point of control of rtorrent.

I was reverse engineering the work flow below with pyrosimple when this problem happened.  

https://pyrocore.readthedocs.io/en/latest/howto.html#host-migration-of-data-state

But now I really cannot do much other than just get all my data + the contents of .session.   Can some person suggest a possible workflow that will allow me to hopefully reseed all the content that I was seeding in my old but broken box but now in the new box which is in a good place now.  

Thank you!

**Comment by kannibalox** [comment]
First things first, throughout this entire process you should always keep an untouched copy of you session files around in case you ever need to start from scratch.

* Copy all the files from your old session to your new session directory (if `rtorrent.lock` exists, you can delete it)
* `cd` to the new session directory and run `chtor -V -s state=+0 *.rtorrent` to set them all to stopped (it'll complain about a bad announce URL, just ignore that)
* Start up the new rtorrent

At this point you can run `rtcontrol` like normal, and you have a lot more options. If all the files are in the exact same spot (i.e. not just the same structure), you shouldn't have to do anything more. If they're in the same structure, but a different parent directory, `rtcontrol --exec d.directory.set=/mnt/new/dir //` will update them all. If they're all over the place, you may have to break it out into several runs, e.g. `rtcontrol --exec d.directory.set=/mnt/new/dir path=/mnt/old/dir/\*`

Once you think you have all the paths set correctly, do a hash check on a couple to ensure things are set correctly, before starting all the rest.

**Comment by govardha** [comment]
Hello Kannibalox,

Thanks for your response, I am attempting to try this on a single torrent, one minor correction in step 2 should refer to the $i.torrent I think.  

I will try this out over the next few days.

**Reply by kannibalox** [reply]
That was on purpose, $i.torrent is a direct copy of the original torrent, while $i.torrent.rtorrent holds a bunch of state information.

**Reply by govardha** [reply]
Ok then I may be doing something wrong or there is some other issue I am running into.  I tried this with multiple files and I get the same error.  

```
$ chtor -V -s state=+0 <hash>.torrent.rtorrent
WARNING:pyrosimple.scripts.chtor.MetafileChanger:Metafile '<hash>.torrent.rtorrent' failed integrity check: bad announce URL - not a string
```

**Reply by kannibalox** [reply]
That is the "bad announce URL" message that I mentioned it would complain about, you can ignore that (although I see how that being the only output is misleading, I'll probably change it).  `chtor` is mostly used for modifying .torrent files so there are some validation checks, but the `-V` flag allows it to modify any bencoded file with just the warning message.

**Reply by govardha** [reply]
I am sorry that I did not read that note about the "bad announce URL", I totally over looked it.  I will give that a shot.  

Thank you again for this amazing work.

**Reply by govardha** [reply]
Hi Kannibalox, 

This worked fantastic, especially with the newer versions where chtor runs clean.  A few final questions before I close this thread.

I am doing a batch onboard of old sessions in my new box.
1.  When I do a hash-check I get the prompt back right away, but when I look at the ui or console rtorrent is gradually doing its thing.  Is there a command that will show me the hash checks that are being performed if any during a current time?
2. Is it possible for me to poke rtorrent to load new session files that are placed into the $SESSION_DIR rather than start/restart of rtorrent that I am currently doing?

Thanks again!

**Reply by kannibalox** [reply]
1. There's two methods that would be useful here: [d.is_hash_checking](https://rtorrent-docs.readthedocs.io/en/latest/cmd-ref.html#term-d-is-hash-checking) and [d.hashing](https://rtorrent-docs.readthedocs.io/en/latest/cmd-ref.html#term-d-hashing): `rtcontrol -o d_hashing,d_is_hash_checking` NB: despite the output looking like a number, the `d_*` fields are often treated like strings. Here's a quick custom field that makes it a proper number:
```python
    yield engine.DynamicField(
	int,
	"hashing_status",
        "is it actively checking",
	matcher=matching.FloatFilter,
        accessor=lambda o: o.rpc_call("d.hashing"),
	requires=["d.hashing"]
    )
```
3. There is no way to do this unfortunately. The session directory isn't meant to be touched at all while rtorrent is running.
